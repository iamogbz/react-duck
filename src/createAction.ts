export function createAction<
    Type extends string = string,
    Payload = unknown,
    State = unknown,
>(
    type: Type,
    prepare = (payload?: Payload): Partial<Action<Type, Payload>> => ({
        payload,
    }),
): ActionCreator<Type, Payload, State> {
    return function createAction(payload?: Payload) {
        const partialAction = prepare(payload);
        return { ...partialAction, type };
    };
}
