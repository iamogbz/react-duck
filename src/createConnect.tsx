import React from "react";

export function createConnect<State>(
    Context: React.Context<ContextValue<State>>,
) {
    return function connect<R, S = R, D = R>(
        mapStateToProps?: MapStateToProps<State, R, S>,
        mapDispatchToProps?: MapDispatchToProps<ContextDispatch, R, D>,
    ) {
        return function wrapWithConnect(
            WrappedComponent:
                | React.FunctionComponent<R>
                | React.ComponentClass<R, State>,
        ): React.FunctionComponent<R> {
            const wrappedComponentName =
                WrappedComponent.displayName ||
                WrappedComponent.name ||
                "Component";
            const displayName = `Connect(${wrappedComponentName})`;

            function WrapperComponent(props: R): React.ReactElement {
                const { state, dispatch } = React.useContext(Context);
                const mappedStateProps = React.useMemo(
                    () => mapStateToProps?.(state, props),
                    [state, props],
                );
                const mappedDispatchToProps = React.useMemo(
                    () => mapDispatchToProps?.(dispatch, props),
                    [dispatch, props],
                );
                return (
                    <WrappedComponent
                        {...props}
                        {...mappedStateProps}
                        {...mappedDispatchToProps}
                    />
                );
            }

            WrapperComponent.WrappedComponent = WrappedComponent;
            WrapperComponent.displayName = displayName;
            return WrapperComponent;
        };
    };
}
