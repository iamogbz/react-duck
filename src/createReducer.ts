export function createReducer<State>(
    initialState: State,
    actionHandlers: ActionReducerMapping<State>,
    defaultHandler: Reducer<State> = (s): State => s,
): Reducer<State> {
    return (state = initialState, action?): State => {
        for (const [actionType, reducer] of Object.entries(actionHandlers)) {
            if (actionType === action?.type) {
                return reducer(state, action);
            }
        }
        return defaultHandler(state, action);
    };
}
