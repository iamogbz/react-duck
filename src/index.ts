export { createAction } from "./createAction";
export { createDuck } from "./createDuck";
export { createReducer } from "./createReducer";
export { createRootReducer } from "./createRootReducer";
