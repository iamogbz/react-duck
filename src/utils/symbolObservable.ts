declare global {
    interface SymbolConstructor {
        readonly observable: symbol;
    }
}

export const SymbolObservable: SymbolConstructor["observable"] =
    Symbol.observable || Symbol("@@observable");
