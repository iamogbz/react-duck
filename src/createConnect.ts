import { connect } from "./utils/connect";

export function createConnect<S, I, T extends string, P, K>(
    Context?: Context<S, T, P>,
) {
    return function connectContext(
        mapStateToProps?: MapStateToProps<S, I, K>,
        mapDispatchToProps?:
            | MapDispatchToProps<T, P, I>
            | ActionCreatorMapping<T, P, S>,
        mergeProps?: (
            stateProps?: K,
            dispatchProps?: ActionDispatcherMapping<T, P>,
            ownProps?: I,
        ) => I & K & ActionDispatcherMapping<T, P>,
        options?: ConnectOptions<S, T, P, I, K>,
    ): (
        component: ReactComponent<I & K & ActionDispatcherMapping<T, P>>,
    ) => React.FunctionComponent<I> {
        return connect(mapStateToProps, mapDispatchToProps, mergeProps, {
            ...options,
            context: Context,
        });
    };
}
