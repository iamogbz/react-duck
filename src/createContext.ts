import * as React from "react";
import { SymbolObservable } from "./utils/symbolObservable";

function createUnimplemented(objectName?: string): (m: string) => () => never {
    const prefix = objectName ? `${objectName}.` : "";
    return function unimplemented(methodName) {
        return (): never => {
            throw new Error(`Unimplemented method: ${prefix}${methodName}`);
        };
    };
}

const idFn = <A>(a: A): A => a;

export function createContext<S, T extends string, P>(
    rootReducer: Reducer<S, T, P>,
    preloadedState: S,
    displayName?: string,
): Context<S, T, P> {
    const unimplemented = createUnimplemented(`Context(${displayName ?? ""})`);
    const Context = React.createContext<ContextValue<S, T, P>>({
        dispatch: idFn,
        enhance: idFn,
        getState: () => preloadedState,
        reducer: rootReducer,
        state: preloadedState,
        [SymbolObservable]: unimplemented(SymbolObservable.toString()),
    });
    if (displayName) {
        Context.displayName = displayName;
    }
    return Context;
}
