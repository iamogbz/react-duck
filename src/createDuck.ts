import { createReducer } from "./createReducer";
import { createAction } from "./createAction";
import { getEntries } from "./utils/getEntries";
import { combineSelectors } from "./utils/combineSelectors";

function getNS<N extends string, T extends string, V extends string>(
    name: N,
    actionType: T,
): V {
    return `${name}/${actionType}` as V;
}

export function createDuck<
    S,
    P,
    R,
    T extends string,
    U extends string,
    N extends string,
    Q extends string
>({
    name,
    initialState,
    reducers,
    selectors,
    actionMapping,
}: {
    name: N;
    initialState: S;
    reducers: ActionReducerMapping<S, U, P>; // Reducers = { CreatorName: Reducer(State, ActionType, PayloadTypes) }
    selectors?: SelectorMapping<S, R, T, P, Q>;
    actionMapping?: Record<T, U>; // ActionMapping = { MappedActionType: ActionType }
}): Duck<S, N, T, P, R, Q, U> {
    const mappedActionTypes = { ...actionMapping } as Record<T, U>;
    const actions = {} as ActionCreatorMapping<T, P, S, U>;
    for (const [actionType] of getEntries(reducers)) {
        const namespacedActionType = getNS<N, U, T>(name, actionType);
        actions[actionType] = createAction(namespacedActionType);
        mappedActionTypes[namespacedActionType] = actionType;
    }

    const actionReducer = createReducer<S, P, U>(initialState, reducers);
    const isMappedActionType = (a?: Action<string, P>): a is Action<T, P> =>
        a !== undefined &&
        Object.prototype.hasOwnProperty.call(mappedActionTypes, a.type);
    function reducer(state: S, action: Action<T, P> | Action<U, P>): S {
        return actionReducer(state, {
            ...action,
            type: isMappedActionType(action)
                ? mappedActionTypes[action.type]
                : action.type,
        });
    }

    return {
        actions,
        initialState,
        name,
        reducer,
        selectors: combineSelectors(name, selectors),
    };
}
