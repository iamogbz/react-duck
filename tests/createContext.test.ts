import * as React from "react";
import { renderHook } from "@testing-library/react-hooks";
import { createContext } from "src";
import { SymbolObservable } from "src/utils/symbolObservable";

describe("createContext", () => {
    it("creates context without displayname", () => {
        const Context = createContext((s) => s, {});
        expect(Context.displayName).toBeUndefined;
    });

    it("creates context with displayname", () => {
        const Context = createContext((s) => s, {}, undefined, "TextContext");
        expect(Context.displayName).toEqual("TextContext");
    });

    it("has unimplemented observable symbol", () => {
        const Context = createContext((s) => s, {});
        const { result } = renderHook(() => React.useContext(Context));
        expect(result.current).toMatchSnapshot();
        expect(
            ((result.current as unknown) as Record<string, unknown>)[
                (SymbolObservable as unknown) as string
            ],
        ).toThrow("Unimplemented method");
    });
});
