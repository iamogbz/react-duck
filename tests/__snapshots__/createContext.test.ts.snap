// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContext has unimplemented observable symbol 1`] = `
Object {
  "dispatch": [Function],
  "enhance": [Function],
  "getState": [Function],
  "reducer": [Function],
  "state": Object {},
  Symbol(@@observable): [Function],
}
`;
